File: benchmark/synduce/list/polyhom.f
Config NonLinear = true;
Config ExtraGrammar = "AutoLifter";
Inductive List = nil Unit | cons {Int, List};
Inductive CList = cnil Unit | single Int | concat {CList, CList};

max = \a: Int. \b: Int. if < a b then b else a;

@Input x: Int;

spec = \xs: List. (fix (
  \f: List -> {Int, Int}. \xs: List.
  match xs with
    nil _ -> {0, 1}
  | cons {hd, tl} -> 
    let result = f tl in
    {+ result.1 (* hd result.2), * x result.2}
  end
) xs).1;

cat = fix (
  \f: List -> List -> List. \xs: List. \ys: List.
  match xs with
    nil _ -> ys
  | cons {a, b} -> cons {a, f b ys}
  end
);

repr = fix (
  \f: CList -> Compress List. \xs: CList.
  match xs with
    cnil _ -> nil unit
  | single a -> cons {a, nil unit}
  | concat {a, b} -> cat (f a) (f b)
  end
);

main = \x: CList. spec (repr x);

File: ml-benchmark/synduce/list/polyhom.ml
config NonLinear = true 
config EnableDeepCoder = true
type list = Nil of unit | Cons of int * list
type clist = CNil of unit | Single of int | Concat of clist * clist

let max a b = if a < b then b else a

@Input val x: int 

let spec xs =
  let rec f xs =
    match xs with
    | Nil _ -> (0, 1)
    | Cons (hd, tl) ->
        let result = f tl in
        match result with
        | (r1, r2) -> ((r1 + (hd * r2)), (x * r2))
  in
  match f xs with
  | (v1, v2) -> v1

let rec cat xs ys =
  match xs with
  | Nil _ -> ys
  | Cons (a, b) -> Cons (a, cat b ys)

val repr: clist -> list compress
let rec repr xs =
  match xs with
  | CNil _ -> Nil ()
  | Single a -> Cons (a, Nil ())
  | Concat (a, b) -> cat (repr a) (repr b)

let program c = spec (repr c)